test/e2e/apps/job.go

```diff
@@ -83,6 +82,10 @@ var _ = SIGDescribe("Job", func() {
 		job, err := e2ejob.CreateJob(ctx, f.ClientSet, f.Namespace.Name, job)
 		framework.ExpectNoError(err, "failed to create job in namespace: %s", f.Namespace.Name)
 
+		ginkgo.By("Awaiting for the job to have the interim success condition")
```

Comment:
Is there an existing e2e test that validates that these conditions are not set until all pods are terminal?  Or was there a decision to test it only in unit and integration due to the lower likelihood that we could observe it?

While it tends to be harder to test, we have done it before for pods using a list/watch construct to accumulate the set of deltas and failed if the invariant is violated.  I suppose in this case we can't guarantee that pods and jobs have happens-before relationships between the two resource types across all possible cluster implementations.

Reply:
> Is there an existing e2e test that validates that these conditions are not set until all pods are terminal? 

Not currently, but I'm open to discuss / add such a test.

> Or was there a decision to test it only in unit and integration due to the lower likelihood that we could observe it?

I think it wasn't discussed during the [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/4368-support-managed-by-for-batch-jobs#e2e-tests), so I'm not sure what others think, but I was a bit concerned with the gain / effort ratio. I think we don't really aim to test kubelet itself with Job test, so integrations tests are fairly good representative of what would happen. In particular, one needs to be careful about the ordering of events. I can imagine that when all pods transition to terminal it is technically possible, that the Job controller informer receives the event before test code receives it. As a result, there is a possibility that Job controller marks the Job failed, while the testing code still sees some of the pods not terminal. (didn't check this is a real possibility, sounds unlikely, but I don't see what can prevent it in principle other than timing).

What is relatively easy to do in e2e tests is that Job's `status.terminating=0` when the job is terminal, which is a Job-level proxy for all pods being terminal. I checked this manually by running the validation in https://github.com/kubernetes/kubernetes/pull/125849.

Reply:
Agree the more complex test is necessary.

Verifying that terminating = 0 when the job final status is set would accomplish the objective of ensuring this behavior is both verified in conformant implementations and prevent regressions.  Is that captured in an existing e2e test or can we add it in this e2e?

Reply:
Not captured currently, but easy to add. I don't think we need a dedicated new test for it. I think adding the check to one or a couple of existing e2e tests should be enough. WDYT?

Reply:
Just added tentatively the asserts: https://github.com/kubernetes/kubernetes/pull/125510/commits/2da2618ca8d52b453f06415fec4774e7df924431. Two remarks:
1. to enable the asserts for successful Jobs (status.succeeded = spec.completions), I also now check the condition is Complete (otherwise there could be race conditions).
2. I don't add the asserts for conformance tests, because we only run them for GA feature gates, and the fix is behind beta. eventually, we will be able to verify all the checks in the helper function for Complete / Failed Jobs.

Reply:
2 is fine.

Reply:
Updated accordingly in the last commit